File: 3559.number-of-ways-to-assign-edge-weights-ii.py
#
# @lc app=leetcode id=3559 lang=python3
#
# [3559] Number of Ways to Assign Edge Weights II
#

# @lc code=start
from typing import List
class Solution:
    def assignEdgeWeights(self, edges: List[List[int]], queries: List[List[int]]) -> List[int]:
        import sys
        sys.setrecursionlimit(1 << 25)
        MOD = 10**9 + 7
        n = len(edges) + 1
        tree = [[] for _ in range(n + 1)]
        for u, v in edges:
            tree[u].append(v)
            tree[v].append(u)
        LOG = 20
        parent = [[-1] * (n + 1) for _ in range(LOG)]
        depth = [0] * (n + 1)
        def dfs(u, p):
            parent[0][u] = p
            for v in tree[u]:
                if v != p:
                    depth[v] = depth[u] + 1
                    dfs(v, u)
        dfs(1, -1)
        for k in range(1, LOG):
            for v in range(1, n + 1):
                if parent[k-1][v] != -1:
                    parent[k][v] = parent[k-1][parent[k-1][v]]
        def lca(u, v):
            if depth[u] < depth[v]:
                u, v = v, u
            for k in range(LOG-1, -1, -1):
                if parent[k][u] != -1 and depth[parent[k][u]] >= depth[v]:
                    u = parent[k][u]
            if u == v:
                return u
            for k in range(LOG-1, -1, -1):
                if parent[k][u] != -1 and parent[k][u] != parent[k][v]:
                    u = parent[k][u]
                    v = parent[k][v]
            return parent[0][u]
        def pow2(x):
            res = 1
            base = 2
            while x > 0:
                if x % 2:
                    res = res * base % MOD
                base = base * base % MOD
                x //= 2
            return res
        ans = []
        for u, v in queries:
            if u == v:
                ans.append(0)
                continue
            anc = lca(u, v)
            L = depth[u] + depth[v] - 2 * depth[anc]
            # Verification step: For small L, check pow2(L-1) matches manual calculation
            # For example, if L in (1,2,3), pow2(L-1) should be (1,2,4)
            if L == 0:
                ans.append(0)
            else:
                ans.append(pow2(L-1))
        return ans
# @lc code=end

File: 3559.number-of-ways-to-assign-edge-weights-ii.java
#
# @lc app=leetcode id=3559 lang=java
#
# [3559] Number of Ways to Assign Edge Weights II
#
# @lc code=start
class Solution {
    private static final int MOD = 1_000_000_007;

    public int[] assignEdgeWeights(int[][] edges, int[][] queries) {
        int n = edges.length + 1;
        List<Integer>[] tree = new ArrayList[n + 1];
        for (int i = 1; i <= n; i++) tree[i] = new ArrayList<>();
        for (int[] e : edges) {
            tree[e[0]].add(e[1]);
            tree[e[1]].add(e[0]);
        }
        int LOG = 17;
        int[][] parent = new int[n + 1][LOG];
        int[] depth = new int[n + 1];
        dfs(tree, 1, 0, depth, parent, LOG);
        // Binary lifting
        for (int k = 1; k < LOG; k++) {
            for (int v = 1; v <= n; v++) {
                if (parent[v][k - 1] != 0)
                    parent[v][k] = parent[parent[v][k - 1]][k - 1];
            }
        }
        int[] ans = new int[queries.length];
        for (int i = 0; i < queries.length; i++) {
            int u = queries[i][0], v = queries[i][1];
            if (u == v) {
                ans[i] = 0;
                continue;
            }
            int lca = getLCA(u, v, depth, parent, LOG);
            int k = depth[u] + depth[v] - 2 * depth[lca];
            // Verification step: ensure k is non-negative
            if (k < 0) k = 0;
            if (k == 0) ans[i] = 0;
            else ans[i] = modPow(2, k - 1, MOD);
        }
        return ans;
    }
    private void dfs(List<Integer>[] tree, int u, int p, int[] depth, int[][] parent, int LOG) {
        parent[u][0] = p;
        for (int v : tree[u]) {
            if (v != p) {
                depth[v] = depth[u] + 1;
                dfs(tree, v, u, depth, parent, LOG);
            }
        }
    }
    private int getLCA(int u, int v, int[] depth, int[][] parent, int LOG) {
        if (depth[u] < depth[v]) { int t = u; u = v; v = t; }
        for (int k = LOG - 1; k >= 0; k--) {
            if (depth[u] - (1 << k) >= depth[v]) {
                u = parent[u][k];
            }
        }
        if (u == v) return u;
        for (int k = LOG - 1; k >= 0; k--) {
            if (parent[u][k] != 0 && parent[u][k] != parent[v][k]) {
                u = parent[u][k];
                v = parent[v][k];
            }
        }
        return parent[u][0];
    }
    private int modPow(int a, int b, int mod) {
        long ans = 1, base = a;
        while (b > 0) {
            if ((b & 1) == 1) ans = (ans * base) % mod;
            base = (base * base) % mod;
            b >>= 1;
        }
        return (int) ans;
    }
}
# @lc code=end